Control/lib/control-core/GrpcServiceClient.js

```
@@ -0,0 +1,297 @@
+   * https://grpc.io/grpc/node/grpc.Client.html
+   * @param {object} config - Configuration for the gRPC client.
+   * @param {string} config.hostname - The hostname of the gRPC server.
+   * @param {number} config.port - The port of the gRPC server.
+   * @param {string} config.label - The service label defined in the proto file.
+   * @param {string} config.package - The package name defined in the proto file.
+   * @param {number} [config.timeout=30000] - Timeout for gRPC calls in milliseconds.
+   * @param {number} [config.connectionTimeout=10000] - Timeout for establishing a connection in milliseconds.
+   * @param {number} [config.maxMessageLength=50] - Maximum message size allowed in MB.
+   * @param {string} protoPath - Path to the proto file defining the gRPC service.
+   */
+  constructor(config, path) {
+    this._logger = LogManager.getLogger(`${process.env.npm_config_log_label ?? 'cog'}/GrpcServiceClient`);
+
+    this._isConnectionReady = false;
+    this._connectionError = null;
+
+    if (this._isConfigurationValid(config, path)) {
+      this._label = config.label;
+      this._package = config.package;
+      this._timeout = config.timeout ?? 30000;
+      this._connectionTimeout = config.connectionTimeout ?? 10000;
+      this._retryConnectionInterval = this._connectionTimeout + 1000; 
+      this._maxMessageLength = config.maxMessageLength ?? 50;
+  
+      const address = `${config.hostname}:${config.port}`;
+      const packageDefinition = protoLoader.loadSync(path, {longs: String, keepCase: true, arrays: true});
+      const octlProto = grpcLibrary.loadPackageDefinition(packageDefinition);
+      const protoService = octlProto[this._package][this._label];
+  
+      const credentials = grpcLibrary.credentials.createInsecure();
+      const options = {'grpc.max_receive_message_length': 1024 * 1024 * this._maxMessageLength}; // MB
+
+      this.client = new protoService(address, credentials, options);
```

handle error instantiating (try catch) or not existing protoService

---

Could you please explain what you are referring to?

---

if there is any error trying to instantiate the service, it is not caught

---

As far as I know, it is not good practice for constructors to do that and instead is better to let the error propagate to the caller